### PR TITLE
Support `Message` in the balances API

### PR DESCRIPTION
**Pull request by xgreenx**
Added support of messages and updated tests. But the test fails because `fuel-tx` doesn't support it. Fix: https://github.com/FuelLabs/fuel-tx/pull/186

Closes https://github.com/FuelLabs/fuel-core/issues/614

### CODE REVIEW

**Review by ControlCplusControlV — COMMENTED** [review]
LGTM overall, just some minor nits/questions

File: fuel-core/src/schema/balance.rs [hunk 1]
```diff
@@ -75,37 +78,32 @@ impl BalanceQuery {
         #[graphql(desc = "asset_id of the coin")] asset_id: AssetId,
     ) -> async_graphql::Result<Balance> {
         let db = ctx.data_unchecked::<Database>();
+        let owner = owner.into();
+        let asset_id = asset_id.into();
+
+        let balance = AssetQuery::new(
+            &owner,
+            &AssetSpendTarget::new(asset_id, u64::MAX, u64::MAX),
+            None,
```
[the remark this answers is not in the document]
Comment (xgreenx) [reply fuel-core/src/schema/balance.rs:87]: Exclude field already tested in `resources.rs`. I reused `AssetQuery` created for resources here=)

File: fuel-tests/tests/balances.rs [hunk 2]
```diff
@@ -122,6 +143,7 @@ async fn balance() {
         .add_witness(Default::default())
         .finalize();
 
+    // TODO: Doesn't work for `MessageSigned` because `fuel-tx` doesn't support it
```
[the remark this answers is not in the document]
Comment (xgreenx) [reply fuel-tests/tests/balances.rs:146]: Yep, I forgot to remove this comment=) https://github.com/FuelLabs/fuel-tx/pull/186

**Review by ControlCplusControlV — DISMISSED** [review]

File: fuel-core/src/database/resource.rs [hunk 3]
```diff
@@ -126,8 +127,13 @@ impl<'a> AssetQuery<'a> {
...
                 if let Resource::Coin { fields, .. } = coin {
-                    fields.asset_id == self.asset.id
-                        && fields.status == CoinStatus::Unspent
+                    let allowed = if let Some(assets) = &self.assets {
+                        assets.contains(&fields.asset_id)
+                    } else {
+                        true
+                    };
+
+                    allowed && fields.status == CoinStatus::Unspent
```
Comment (leviathanbeak) [inline fuel-core/src/database/resource.rs:136]: I'd turn this upside down a bit:

```suggestion
    let is_unspent = fields.status == CoinStatus::Unspent;

    self.assets
        .as_ref()
        .map(|assets| assets.contains(&fields.asset_id) && is_unspent)
        .unwrap_or(is_unspent)
```

File: fuel-core/src/database/resource.rs [hunk 4]
```diff
@@ -166,8 +172,48 @@ impl<'a> AssetQuery<'a> {
...
             });
 
-        coins_iter
-            .chain(messages_iter.take_while(|_| self.asset.id == ChainConfig::BASE_ASSET))
+        coins_iter.chain(messages_iter.take_while(|_| {
+            if let Some(assets) = &self.assets {
+                assets.contains(&ChainConfig::BASE_ASSET)
+            } else {
+                true
+            }
```
Comment (leviathanbeak) [inline fuel-core/src/database/resource.rs:180]: Same here:

```suggestion
    self.assets
        .as_ref()
        .map(|assets| assets.contains(&ChainConfig::BASE_ASSET))
        .unwrap_or(true)
```

**Comment by leviathanbeak** [comment]
Looks good, but I'd wait for `fuel-tx` issue to be fixed and updated so that we don't have to patch this one with another branch

**Comment by xgreenx** [comment]
Blocked by https://github.com/FuelLabs/fuel-tx/pull/186

**Review by leviathanbeak — APPROVED** [review]
:shipit:

**Pull request merged**